Namespace all components

cdrage opened the pull request:
Closes https://github.com/redhat-developer/ocdev/issues/214

Allows the ability to define the same component name but on different applications.

cdrage commented:
Hey @kadel you said that there is a lot of code to change in order to fix this, but I managed to get it work with just this little amount of change.

[WIP] for now since I haven't updated ImageStream yet:
```
Errors:
  * bc/foobar-nodejs is pushing to istag/nodejs:latest, but the image stream for that tag does not exist.
  * bc/nodeapp-nodejs is pushing to istag/nodejs:latest, but the image stream for that tag does not exist.
  * The image trigger for dc/foobar-nodejs will have no effect because is/nodejs does not exist.
  * The image trigger for dc/nodeapp-nodejs will have no effect because is/nodejs does not exist.
```

cdrage commented:
Unfortunately I'm unable to diagnose what's failing here... There's zero error output within the e2e tests. @ashetty1 

```sh
make test-e2e
go test github.com/redhat-developer/ocdev/tests/e2e

Running Suite: ocdev test suite
===============================
Random Seed: 1522868812
Will run 9 of 9 specs

New project created and now using project : ocdev-1522868812

ocdev-1522868812

•Creating application: usecase5
Switched to application: usecase5

usecase5
•ocdev-1522868812

•Creating application: usecase5-2
Switched to application: usecase5-2

usecase5-2
Switched to application: usecase5

usecase5
•exit status 1:
FAIL    github.com/redhat-developer/ocdev/tests/e2e     9.436s
Makefile:72: recipe for target 'test-e2e' failed
make: *** [test-e2e] Error 1
```

When I search for "usecase5" within the test_e2e.go file I get multiple functions..

kadel commented:
I haven't tested this, but I suspect that this will break `ocdev update` and `ocdev push` commands.
You need to update those too.
This might also be a reason why tests is failing.

kadel commented on pkg/occlient/occlient.go:
```diff
@@ -444,7 +444,11 @@ func (c *Client) GetImageStreamsNames(namespace string) ([]string, error) {
 // NewAppS2I create new application using S2I
 // if gitUrl is ""  than it creates binary build otherwise uses gitUrl as buildSource
-func (c *Client) NewAppS2I(name string, builderImage string, gitUrl string, labels map[string]string, annotations map[string]string) (string, error) {
+func (c *Client) NewAppS2I(applicationName, componentName string, builderImage string, gitUrl string, labels map[string]string, annotations map[string]string) (string, error) {
```
anything `occlient` shouldn't have any references to component or application. We should have clean abstraction hierarchy.

cdrage replied:
No worries, I'll change this up to be higher up in the code :+1: you're right. It shouldn't have any references to neither of those (especially since we're just combining the strings)

kadel commented on pkg/component/component.go:
```diff
@@ -66,7 +66,7 @@ func CreateFromGit(client *occlient.Client, name string, ctype string, url strin
 	annotations := map[string]string{componentSourceURLAnnotation: url}
 
-	output, err := client.NewAppS2I(name, ctype, url, labels, annotations)
+	output, err := client.NewAppS2I(currentApplication, name, ctype, url, labels, annotations)
```
`name` argument in `NewAppS2I` is not a name of the component. It is name for OpenShift objects, we have been just passing the same value as component name till now.

You can pass in new namespaced name here without the need to add a new argument to that function.

The component name is determined by label (`GetLabels`)

cdrage replied:
I'll refactor this :+1: thx!

cdrage commented:
@ashetty1 Is it possible for you to build this (then do a `cp ocdev` binary to your folder) and try this out with the end-to-end testing?

I'm trying to make this pass but I'm getting stuck on the "pinging" code.

```sh
Usecase #5 When push is made                                                                      
  should push reflect changes                                                                      
  /home/wikus/seafile/files/dev/go/src/github.com/redhat-developer/ocdev/tests/e2e/e2e_test.go:195
$ oc get svc usecase5-nodejs -o go-template='{{.spec.clusterIP}}:{{(index .spec.ports 0).port}}'
172.30.79.251:8080                                   
                                                                                                                                        
Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging h$
tp://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.$
0.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:$
080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080Pinging http://172.30.79.251:8080
• Failure [309.195 seconds]                                                     
Usecase #5                                                                                        
/home/wikus/seafile/files/dev/go/src/github.com/redhat-developer/ocdev/tests/e2e/e2e_test.go:116  
  When push is made                                                                                 
  /home/wikus/seafile/files/dev/go/src/github.com/redhat-developer/ocdev/tests/e2e/e2e_test.go:194
    should push reflect changes [It]
    /home/wikus/seafile/files/dev/go/src/github.com/redhat-developer/ocdev/tests/e2e/e2e_test.go:195
                                                                                                   
    could not ping the specific service in given time: 10 minutes                                 
                                                                                                               
    /home/wikus/seafile/files/dev/go/src/github.com/redhat-developer/ocdev/tests/e2e/e2e_test.go:81                    
------------------------------                                                                             
Usecase #5 adding storage                                   
  should default to active component when no component name is passed
  /home/wikus/seafile/files/dev/go/src/github.com/redhat-developer/ocdev/tests/e2e/e2e_test.go:240
                                                                                         
```

Ignore the "Ping http://172.30.79.251:8080" I added that to have a bit of verbosity. Does this pass locally on your machine?

cdrage commented:
@ashetty1 

So I looked into it further... 

Tests seem to pass now, but that's only because I uncommented the 

I don't see why we are pinging the cluster IP despite nothing appearing on `8080`. There's no `oc expose` or `ocdev url` before the ping.

How are we using: https://github.com/redhat-developer/ocdev/blob/master/tests/e2e/e2e_test.go#L199 for pinging when only ClusterIP is exposed (and I'm unable to ping that endpoint via minishift)

ashetty1 commented:
Sorry, missed this. Will look into this earliest tomorrow.

Possibly, https://github.com/redhat-developer/ocdev/blob/d69a341345aaadf56ba9fdbeefeac49419b9442d/tests/e2e/e2e_test.go#L196 won't work for all cases?

cdrage commented:
@ashetty1 That's what I was thinking. I removed https://github.com/redhat-developer/ocdev/blob/master/tests/e2e/e2e_test.go#L196 from the testing suite and this is able to pass (with the exception of storage which I'm working on)

cdrage commented:
@ashetty1 

A few things testing side that I had to modify that were not working on `minishift` that I've added to this PR:

 - Removed the "ping" function
 - Removed running at `oc` `system:admin` (it is actually not needed and tests pass fine without it

cdrage commented:
Ayyyy tests are green :100:

ashetty1 commented on tests/e2e/e2e_test.go:
```diff
@@ -193,14 +195,14 @@ var _ = Describe("Usecase #5", func() {
+			// " | grep 'Welcome to your Node.js application on OpenShift'")
 
-			log.Printf("Text before change: %s", strings.TrimSpace(grepBeforePush))
+			// log.Printf("Text before change: %s", strings.TrimSpace(grepBeforePush))
```
By commenting out this section, we won't be doing a `curl` before updating the html file. So there won't be any before and after verification. Are we okay with that?

cdrage replied:
@ashetty1 There still is, we run sed before pushing: `https://github.com/redhat-developer/ocdev/pull/303/files/f626589e6eff5cc19b89453adc6ab26581247275#diff-73bf4cf8a09657b595c016b77c1d805bR208`

cdrage replied:
the problem is that pinging doesn't work using `minishift` since the IP address is private and inaccessible.

ashetty1 replied:
@cdrage ah! That's a good enough reason I guess :+1:

Might make sense to add a note there? I will see if we can do it in any other way.

ashetty1 commented:
@cdrage I have addressed the admin login part while refactoring teh e2e code: #322. Thank you so much for that tip!

cdrage commented:
@kadel @ashetty1 @containscafeine ready for review :+1:

ashetty1 commented on tests/e2e/e2e_test.go:
```diff
@@ -241,7 +241,7 @@ var _ = Describe("ocdev", func() {
 
 				// Check against path and name against dc
-				getDc := runCmd("oc get dc/nodejs -o go-template='" +
+				getDc := runCmd("oc get dc/usecase5nodejs -o go-template='" +
```
usecase5-nodejs, no?

cdrage commented:
Blocked until we get @containscafeine 's storage go-client PR in. https://github.com/redhat-developer/ocdev/pull/295

cdrage commented:
https://github.com/redhat-developer/ocdev/pull/295 merged, I'll start work again on this / get this updated.
